<launch>

    <set_env name="RCUTILS_CONSOLE_OUTPUT_FORMAT" value="{function_name}(): {message} "/>
    <set_env name="RCUTILS_COLORIZED_OUTPUT" value="1"/>
    
    <!-- 変数 -->
    <arg name="name_space" default="ns"/> <!-- config/{}.yaml内のnamespaceとそろえる必要あり -->
    <arg name="name_pkg" default="dynamixel_handler"/>

    <!-- 通信設定 -->
    <arg name="baudrate" default="57200"/>
    <arg name="device_name" default="/dev/ttyUSB0"/>
    <arg name="latency_timer" default="16"/>

    <node pkg="$(var name_pkg)" exec="dynamixel_handler_node" name="dyn_handler" namespace="$(var name_space)" output="screen">
        <!-- 通信の設定 -->
        <param name="baudrate" value="$(var baudrate)"/>
        <param name="device_name" value="$(var device_name)"/>
        <param name="latency_timer" value="$(var latency_timer)"/>

        <!-- 細かいパラメータの設定 -->
        <param from="$(find-pkg-share $(var name_pkg))/config/dynamixel_handler.yaml"/>

        <!-- トピックの設定 -->
        <!-- 書き込み -->
        <remap from="dynamixel/commnad"   to="dynamixel/commnad"/>
        <remap from="dynamixel/cmd/x/position"          to="dynamixel/cmd/x/position"/>
        <remap from="dynamixel/cmd/x/velocity"          to="dynamixel/cmd/x/velocity"/>
        <remap from="dynamixel/cmd/x/current"           to="dynamixel/cmd/x/current"/>
        <remap from="dynamixel/cmd/x/current_position"  to="dynamixel/cmd/x/current_position"/>
        <remap from="dynamixel/cmd/x/extended_position" to="dynamixel/cmd/x/extended_position"/>
        <remap from="dynamixel/cmd/p/position"          to="dynamixel/cmd/p/position"/>
        <remap from="dynamixel/cmd/p/velocity"          to="dynamixel/cmd/p/velocity"/>
        <remap from="dynamixel/cmd/p/current"           to="dynamixel/cmd/p/current"/>
        <remap from="dynamixel/cmd/p/extended_position" to="dynamixel/cmd/p/extended_position"/>
        <remap from="dynamixel/opt/limit/w" to="dynamixel/opt/limit/w"/>
        <remap from="dynamixel/opt/gain/w"  to="dynamixel/opt/gain/w" />
        <remap from="dynamixel/opt/mode/w"  to="dynamixel/opt/mode/w" />

        <!-- 読み込み -->
        <remap from="dynamixel/state" to="dynamixel/state"/>
        <remap from="dynamixel/error" to="dynamixel/error"/>
        <remap from="dynamixel/opt/limit/r" to="dynamixel/opt/limit/r"/>
        <remap from="dynamixel/opt/gain/r"  to="dynamixel/opt/gain/r" />
        <remap from="dynamixel/opt/mode/r"  to="dynamixel/opt/mode/r" />
    </node>

</launch>